### PR TITLE
Query: Introduce method for providing brackets around SqlExpression b…

### PR DESCRIPTION
**Pull request by smitpatel**
…ased on operator precedence

This is just base infra for changes. Making changes in the logic and updating SQL in different commit for ease of reviewing

### CODE REVIEW

File: src/EFCore.Relational/Query/QuerySqlGenerator.cs [hunk 1]
```diff
@@ -430,7 +430,7 @@ protected virtual void CheckComposableSql(string sql)
                 {
                     var i = span.IndexOf('\n');
                     span = i > 0
-                        ? span.Slice(i + 1).TrimStart()
+                        ? span[(i + 1)..].TrimStart()
```
Comment (roji) [inline src/EFCore.Relational/Query/QuerySqlGenerator.cs:433]: I like ranges, but in this very particular case I find it less readable...

File: src/EFCore.Relational/Query/QuerySqlGenerator.cs [hunk 2]
```diff
@@ -459,18 +459,11 @@ protected virtual void CheckComposableSql(string sql)
...
-            else if (sql.StartsWith("WITH", StringComparison.OrdinalIgnoreCase))
-            {
-                sql = sql.Slice("WITH".Length);
-            }
-            else
-            {
-                throw new InvalidOperationException(RelationalStrings.FromSqlNonComposable);
-            }
+            sql = sql.StartsWith("SELECT", StringComparison.OrdinalIgnoreCase)
+                ? sql["SELECT".Length..]
```
Comment (roji) [inline src/EFCore.Relational/Query/QuerySqlGenerator.cs:463]: This is nice though :)

File: src/EFCore.Relational/Query/QuerySqlGenerator.cs [hunk 3]
```diff
@@ -790,6 +775,56 @@ protected override Expression VisitIn(InExpression inExpression)
...
             => _operatorMap[binaryExpression.OperatorType];
 
+        /// <summary>
+        ///     Returns a bool value indicating if the inner SQL expression required to be put inside parenthesis
+        ///     when generating SQL for outer SQL expression.
+        /// </summary>
+        /// <param name="outerExpression">The outer expression which provides context in which SQL is being generated.</param>
+        /// <param name="innerExpression">The inner expression which may need to be put inside parenthesis.</param>
+        /// <returns>A bool value indicating that parenthesis is required or not. </returns>
+        protected virtual bool RequiresParenthesis(SqlExpression outerExpression, SqlExpression innerExpression)
```
Comment (roji) [inline src/EFCore.Relational/Query/QuerySqlGenerator.cs:785]: nit: parentheses is the plural of parenthesis (there are two)

```suggestion
        protected virtual bool RequiresParentheses(SqlExpression outerExpression, SqlExpression innerExpression)
```

File: src/EFCore.Relational/Query/QuerySqlGenerator.cs [hunk 4]
```diff
@@ -790,6 +775,56 @@ protected override Expression VisitIn(InExpression inExpression)
...
+        {
+            switch (innerExpression)
```
Comment (roji) [inline src/EFCore.Relational/Query/QuerySqlGenerator.cs:787]: nit: consider making this an expression if that turns out possible in subsequent work

**Review by roji — APPROVED** [review]
LGTM for now - looking forward to seeing the real stuff next...

File: src/EFCore.Relational/Query/QuerySqlGenerator.cs (hunk 4 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/QuerySqlGenerator.cs:787]: The commented out code gets uncommented in very first iteration of removal so won't last long.

File: src/EFCore.Relational/Query/QuerySqlGenerator.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/QuerySqlGenerator.cs:785]: I considered it and left it for API review. Will update name.

File: src/EFCore.Relational/Query/QuerySqlGenerator.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/QuerySqlGenerator.cs:433]: Why though? How is it different from any other range operation?

Comment (roji) [reply src/EFCore.Relational/Query/QuerySqlGenerator.cs:433]: Just compared to `span.Slice(i+1)`. Less parentheses/brackets, mostly. Not important.

Comment (smitpatel) [reply src/EFCore.Relational/Query/QuerySqlGenerator.cs:433]: Lesser characters though.

Comment (roji) [reply src/EFCore.Relational/Query/QuerySqlGenerator.cs:433]: :shrug:

**Pull request merged**